### PR TITLE
Fix ingested file and direcotry not being sync

### LINKED ISSUE #5287: fsync directory after file ingestion
> Note: Please use Issues only for bug reports. For questions, discussions, feature requests, etc. post to dev group: https://www.facebook.com/groups/rocksdb.dev

### Expected behavior

@petermattis pointed me to this. We should sync the DB directory after ingesting an SST file to ensure that SST file can survive a system crash.

### Actual behavior

The directory sync is not happening (see repro below).

### Steps to reproduce the behavior

I verified with `strace` that there is indeed no sync on the directory after creating a file during ingestion: https://gist.github.com/ajkr/053b51bb97d5a8957136b2a8cf885c71

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Summary:
It it not safe to assume application had sync the SST file before ingest it into DB. Also the directory to put the ingested file needs to be fsync, otherwise the file can be lost. For integrity of RocksDB we need to sync the ingested file and directory before apply the change to manifest.

Also syncing after writing global sequence when write_global_seqno=true was removed in #4172. Adding it back.

Fixes #5287.

Test Plan:
Test ingest file with ldb command and observe fsync/fdatasync in strace output. Tried both move_files=true and move_files=false.
https://gist.github.com/yiwu-arbug/650a4023f57979056d83485fa863bef9

More test suggestions are welcome.

### CODE REVIEW

**Comment by yiwu-arbug** [comment]
@siying @riversand963 IMO this is a critical fix. PTAL. Thanks.

@ajkr If you have suggestion for test that's welcome. Thanks as well.

File: db/external_sst_file_ingestion_job.cc [hunk 1]
```diff
@@ -95,8 +98,18 @@ Status ExternalSstFileIngestionJob::Prepare(
                       f.fd.GetPathId());
     if (ingestion_options_.move_files) {
       status = env_->LinkFile(path_outside_db, path_inside_db);
-      if (status.IsNotSupported() &&
-          ingestion_options_.failed_move_fall_back_to_copy) {
+      if (status.ok()) {
```
Comment (riversand963) [inline db/external_sst_file_ingestion_job.cc]: Should we sync the ingested file in both link/copy case?

Comment (riversand963) [reply db/external_sst_file_ingestion_job.cc]: nvm. It will be synced by 'CopyFile'

File: db/external_sst_file_ingestion_job.cc [hunk 2]
```diff
@@ -557,6 +581,9 @@ Status ExternalSstFileIngestionJob::AssignGlobalSeqnoForIngestedFile(
       std::string seqno_val;
       PutFixed64(&seqno_val, seqno);
       status = rwfile->Write(file_to_ingest->global_seqno_offset, seqno_val);
+      if (status.ok()) {
+        status = SyncIngestedFile(rwfile.get());
```
Comment (yiwu-arbug) [inline db/external_sst_file_ingestion_job.cc]: It seems to me we don't need to sync directory here, but I'm not very sure.

File: db/external_sst_file_ingestion_job.cc [hunk 3]
```diff
@@ -115,6 +129,16 @@ Status ExternalSstFileIngestionJob::Prepare(
       break;
     }
     f.internal_file_path = path_inside_db;
+    ingestion_path_ids.insert(f.fd.GetPathId());
+  }
+
+  if (status.ok()) {
+    for (auto path_id : ingestion_path_ids) {
+      status = directories_->GetDataDir(path_id)->Fsync();
+      if (!status.ok()) {
```
Comment (zhangjinpeng1987) [inline db/external_sst_file_ingestion_job.cc]: Should we print some log here?

**Comment by yiwu-arbug** [comment]
@riversand963 Do you have more comments? thanks!

File: db/external_sst_file_ingestion_job.cc (hunk 2 above)
Comment (riversand963) [reply db/external_sst_file_ingestion_job.cc]: Does this mean we may potentially have to sync a file twice if write_global_seqno is true?

File: db/external_sst_file_ingestion_job.cc [hunk 4]
```diff
@@ -95,8 +98,23 @@ Status ExternalSstFileIngestionJob::Prepare(
...
-          ingestion_options_.failed_move_fall_back_to_copy) {
+      if (status.ok()) {
+        // It is unsafe to assume application had sync the file and file
+        // directory before ingest the file. For integrity of RocksDB we need
+        // to sync the file.
+        std::unique_ptr<WritableFile> file_to_sync;
+        status = env_->ReopenWritableFile(path_inside_db, &file_to_sync,
+                                          env_options_);
+        if (status.ok()) {
+          status = SyncIngestedFile(file_to_sync.get());
```
Comment (riversand963) [inline db/external_sst_file_ingestion_job.cc]: Do we have to sync the file here if `write_global_seq` is true?

Comment (yiwu-arbug) [reply db/external_sst_file_ingestion_job.cc]: Yes, it is syncing the file twice. In `Prepare()` we don't hold db mutex and write thread, so syncing does not block other stuff, while when we update global sequence db mutex is held and write thread is blocked. Unfortunately in `Prepare()` even `write_global_seqno=true` we may end up not writing global seqno (e.g. new global seqno = 0).

Comment (riversand963) [reply db/external_sst_file_ingestion_job.cc]: Make sense.

**Comment by riversand963** [comment]
The AppVeyor error should have been fixed in master, and I believe it's irrelevant to this PR.
I think the feature LGTM, since it does not weaken the guarantee that RocksDB already provides. Do you plan to add unit test using fault injection env? For example, you can test the behavior of the db when syncing the sst fails.

**Comment by yiwu-arbug** [comment]
@riversand963 sure, I'll add the test.

**Comment by yiwu-arbug** [comment]
@riversand963 test added. mind take another look? Thanks.

**Comment by yiwu-arbug** [comment]
I didn't notice I left some debug log and failing CI. Fixed. @riversand963 mind take another look? Thanks.

**Comment by yiwu-arbug** [comment]
@riversand963 kindly ping

**Comment by riversand963** [comment]
I think there is compilation error:
```
In file included from test_util/fault_injection_test_env.cc:14:
./test_util/fault_injection_test_env.h:99:10: error: 'GetRequiredBufferAlignment' overrides a member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]
  size_t GetRequiredBufferAlignment() const {
         ^
./include/rocksdb/env.h:888:18: note: overridden virtual function is here
  virtual size_t GetRequiredBufferAlignment() const { return kDefaultPageSize; }
                 ^
1 error generated.
```

**Comment by yiwu-arbug** [comment]
sorry. fixed.

**Comment by riversand963** [comment]
LGTM

**Pull request closed without merging**